oap-server/server-alarm-plugin/src/main/java/org/apache/skywalking/oap/server/core/alarm/provider/CompositeRuleEvaluator.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.apache.skywalking.oap.server.core.alarm.provider;
+
+import com.google.common.base.Joiner;
+import com.google.common.collect.ImmutableListMultimap;
+import com.google.common.collect.Multimaps;
+import lombok.AllArgsConstructor;
+import org.apache.skywalking.oap.server.core.Const;
+import org.apache.skywalking.oap.server.core.alarm.AlarmMessage;
+import org.apache.skywalking.oap.server.core.alarm.provider.expression.Expression;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Evaluator composite rule using expression eval
+ *
+ * @since 8.2.0
+ */
+@AllArgsConstructor
+public class CompositeRuleEvaluator {
+
+    private Expression expression;
+
+    /**
+     * Evaluator composite rule
+     *
+     * @param compositeAlarmRules compositeRules
+     * @param alarmMessages       triggered alarm messages
+     * @return
+     */
+    public List<AlarmMessage> evaluator(List<CompositeAlarmRule> compositeAlarmRules, List<AlarmMessage> alarmMessages) {
+        final List<AlarmMessage> compositeRuleMessages = new ArrayList<>();
+        ImmutableListMultimap<String, AlarmMessage> messageMap = Multimaps.index(alarmMessages, alarmMessage -> Joiner.on(Const.ID_CONNECTOR).useForNull(Const.EMPTY_STRING).join(alarmMessage.getId0(), alarmMessage.getId1()));
+        for (CompositeAlarmRule compositeAlarmRule : compositeAlarmRules) {
+            String expr = compositeAlarmRule.getExpression();
+            Set<String> dependencyRules = expression.analysisInputs(expr);
+            Map<String, Object> dataContext = new HashMap<>();
+            messageMap.asMap().forEach((key, alarmMessageList) -> {
+                dependencyRules.forEach(ruleName -> dataContext.put(ruleName, false));
+                alarmMessageList.forEach(alarmMessage -> {
+                    if (dependencyRules.contains(alarmMessage.getRuleName())) {
+                        dataContext.put(alarmMessage.getRuleName(), true);
```

Comment:
If `contains`, then `put`?

Reply:
Yes, when the composite rule contains this rulename, then put in the data context for eval the expression

Reply:
done